server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java

```diff
@@ -162,12 +157,120 @@ protected Aggregator createInternal(Aggregator parent, CardinalityUpperBound car
         );
     }
 
+    /**
+     * Create the {@link CollectorSource}, gathering some timing information
+     * if we're profiling.
+     * <p>
+     * When profiling aggregations {@link LeafBucketCollector#collect(int, long)} method
+     * out of the box but our implementation of that method does three things that is
+     * useful to get timing for:
+     * <ul>
+     * <li>Fetch field values from {@code _source}
+     * <li>Analyze the field
+     * <li>Do all the normal {@code terms} agg stuff with its terms
+     * </ul>
+     * <p>
+     * The most convenient way to measure all of these is to time the fetch and all
+     * the normal {@code terms} agg stuff. You can then subtract those timings from
+     * the overall collect time to get the analyze time. You can also get the total
+     * number of terms that we analyzed by looking at the invocation count on the
+     * {@code terms} agg stuff.
+     * <p>
+     * While we're at it we count the number of values we fetch from source.
+     */
+    private CollectorSource createCollectorSource() {
+        Analyzer analyzer = context.getIndexAnalyzer(f -> {
+            throw new IllegalArgumentException("No analyzer configured for field " + f);
+        });
+        if (context.profiling()) {
+            return new SignificantTextCollectorSource(
```

Comment:
There are so many overloaded methods here that I feel like extracting it into an old fashioned static class can greatly increase readability.

Reply:
Sure!